File: python/atelier_bench/rules/label_presence.py
"""label-presence rule (nominal: subclasses RuleContract)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.rule_contract import RuleContract

if TYPE_CHECKING:
    from ..dataset import Dataset


class LabelPresenceRule(RuleContract):
    KEY = "label-presence"

    def __init__(self) -> None:
        self.key = self.KEY

    def evaluate(self, data: "Dataset") -> bool:
        return all(len(part.name) > 0 for part in data.parts)

File: go/atelier/rules/label_presence.go
// Package rules carries the 48 rules: 24 NOMINAL rule structs in this package
// (each with a compile-time assertion against RuleContract and a value
// receiver) plus the 24 STRUCTURAL ad-hoc rule values from rules/adhoc, which
// assert nothing and satisfy the same method set by shape alone.
package rules

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
)

// Compile-time proof against the NOMINAL contract, asserted on the VALUE
// type: LabelPresenceRule's methods take value receivers, so both LabelPresenceRule and
// *LabelPresenceRule satisfy RuleContract.
var _ contracts.RuleContract = LabelPresenceRule{}

// LabelPresenceRule is the label-presence rule.
type LabelPresenceRule struct{}

// Key is the rule identifier.
func (LabelPresenceRule) Key() string {
	return "label-presence"
}

// Evaluate reports whether the rule holds on this dataset.
func (r LabelPresenceRule) Evaluate(data *dataset.Dataset) bool {
	for _, part := range data.Parts {
		if len(part.Name) == 0 {
			return false
		}
	}
	return true
}

// init registers this rule.
func init() {
	Register(LabelPresenceRule{})
}
